bpo-41428: Implementation for PEP 604

maggiemoss opened the pull request:
# PEP 604 - Proposed Implementation
https://www.python.org/dev/peps/pep-0604/

Building off of the comments left on this PR here: https://github.com/pprados/cpython/tree/PEP604

<!-- issue-number: [bpo-41428](https://bugs.python.org/issue41428) -->
https://bugs.python.org/issue41428
<!-- /issue-number -->

maggiemoss committed:
Add new type object for union types

maggiemoss committed:
Add test.

maggiemoss committed:
Add basic tp_* fields to union object

maggiemoss committed:
Squash this commit.

maggiemoss committed:
Squash this commit.

maggiemoss committed:
Add union_traverse method.

maggiemoss committed:
Handle Py_None in union.

maggiemoss committed:
Clean up syntax with Py_None

maggiemoss committed:
Update getattro method.

maggiemoss committed:
Reduce code duplication when adding union types together.

maggiemoss commented on Objects/typeobject.c:
```diff
@@ -3741,6 +3741,76 @@ type_is_gc(PyTypeObject *type)
+
+    PyObject *newUnionType=Py_Union(tuple);
+    Py_DECREF(tuple);
+    return newUnionType;
```
There was a test case in the original Implementation: ()

That looked like:

`self.assertEqual(None | int, typing.Union[int, None])`

(This test currently fails with this implementation). Building off the above example, should it also be true that 
`self.assertEqual(int | bool | str, typing.Union[int, str, bool])`

If so, I think a potential solution would be to sort the arguments when we create the `Py_Union` type. Not sure of the best way to implement that sorting functionality.

gvanrossum replied:
In typing.py this is implemented in `__eq__` by converting the args of each union object to a set. You could presumably do the same thing for the new union type.

maggiemoss committed:
Add __or__ method to _SpecialGenericAlias

maggiemoss committed:
Add support for NewType with unions.

maggiemoss committed:
Remove old code from abstract.c

maggiemoss committed:
Fix warning for incompatible pointer type.

gvanrossum commented on Lib/test/test_isinstance.py:
```diff
@@ -208,6 +209,15 @@ def test_isinstance_abstract(self):
         self.assertEqual(False, isinstance(AbstractChild(), Child))
 
+    # def test_isinstance_with_or_union(self):
+    #     self.assertEqual(True, isinstance(AbstractChild(), AbstractChild | int))
```
I would use `self.assertTrue` and `self.assertFalse`.

gvanrossum commented on Lib/test/test_types.py:
```diff
@@ -598,7 +600,91 @@ def test_method_descriptor_types(self):
+        self.assertEqual(typing.List[int] | None, typing.Union[typing.List[int], None])
+        self.assertEqual(None | typing.List[int], typing.Union[None, typing.List[int]])
+        self.assertEqual(
+            BaseException | \
```
You don't need the `\` for line continuation -- everything is inside the outermost `self.assertEqual(...)`.

gvanrossum commented on Lib/typing.py:
```diff
@@ -578,6 +578,13 @@ def __init__(self, name, *constraints, bound=None,
         if def_mod != 'typing':
             self.__module__ = def_mod
 
+    def __or__(self,right):
```
PEP 8 insists on a space after a comma in such situations.

gvanrossum commented on Lib/typing.py:
```diff
@@ -578,6 +578,13 @@ def __init__(self, name, *constraints, bound=None,
+        return Union[self,right]
+    def __ror__(self,right):
+        return Union[self,right]
+    def __invert__(self):
```
Hm, that's not in the PEP, is it? (I think it might have been at some point, but `~X` is far less the obvious way to spell `Optional[X]` than `X|Y` is the obvious way to spell `Union[X, Y]`.

gvanrossum commented on Objects/abstract.c:
```diff
@@ -910,7 +910,7 @@ binary_op(PyObject *v, PyObject *w, const int op_slot, const char *op_name)
         {
             PyErr_Format(PyExc_TypeError,
-                "unsupported operand type(s) for %.100s: "
+                "unsupported operand type(s) for () %.100s: "
```
Looks like you accidentally changed this.

gvanrossum commented on Objects/abstract.c:
```diff
@@ -2469,6 +2469,7 @@ abstract_issubclass(PyObject *derived, PyObject *cls)
     }
 }
 
+
```
I wouldn't fix minor style nits in existing code that you're not otherwise changing. It would make for a noisy code review.

gvanrossum commented on Objects/abstract.c:
```diff
@@ -2511,7 +2512,7 @@ object_isinstance(PyObject *inst, PyObject *cls)
     else {
         if (!check_class(cls,
-            "isinstance() arg 2 must be a type or tuple of types"))
+            "isinstance() arg 2 must be a type, a tuple of types or union"))
```
```suggestion
            "isinstance() arg 2 must be a type, a tuple of types or a union"))
```

gvanrossum commented on Objects/typeobject.c:
```diff
@@ -3741,6 +3741,91 @@ type_is_gc(PyTypeObject *type)
+static int
+is_new_type(PyObject *obj)
+{
+    if (!PyObject_IsInstance((PyObject *)obj, (PyObject *)&PyFunction_Type)) {
```
```suggestion
    if (!PyObject_IsInstance(obj, (PyObject *)&PyFunction_Type)) {
```
(Also below.)

gvanrossum commented on Objects/typeobject.c:
```diff
@@ -3741,6 +3741,91 @@ type_is_gc(PyTypeObject *type)
+    // Check param is a PyType or GenericAlias
+    if ((param == NULL) || is_not_unionable(param)  || is_not_unionable(self))
+    {
+        PyErr_SetString(PyExc_TypeError, "'type' expected");
```
Shouldn't this return `Py_NotImplemented`? (After INCREF.) To let the other argument give it a try.

gvanrossum commented on Objects/typeobject.c:
```diff
@@ -3741,6 +3741,91 @@ type_is_gc(PyTypeObject *type)
+}
+
+static int
+is_not_unionable (PyObject *obj)
```
```suggestion
is_not_unionable(PyObject *obj)
```
Also consider changing the sense -- having a function `is_unionable()` makes more sense. The callers then just add `!` in front.

gvanrossum commented on Objects/typeobject.c:
```diff
@@ -3741,6 +3741,91 @@ type_is_gc(PyTypeObject *type)
+static PyObject *
+type_or(PyTypeObject* self, PyObject* param) {
+    // Check param is a PyType or GenericAlias
+    if ((param == NULL) || is_not_unionable(param)  || is_not_unionable(self))
```
```suggestion
    if ((param == NULL) || is_not_unionable((PyObject *)param)  || is_not_unionable(self))
```
Fixes the error reported by GitHub Actions below.

gvanrossum commented on Objects/typeobject.c:
```diff
@@ -3741,6 +3741,91 @@ type_is_gc(PyTypeObject *type)
+static int
+is_not_unionable (PyObject *obj)
+{
+    return (obj != Py_None) &
```
```suggestion
    return (obj != Py_None) &&
```

gvanrossum commented on Objects/typeobject.c:
```diff
@@ -3741,6 +3741,91 @@ type_is_gc(PyTypeObject *type)
+    } else {
+        PyObject *param_type = param == Py_None ? (PyObject *)Py_TYPE(param) : param;
+        PyTypeObject *self_type = (PyObject *)self == Py_None ? Py_TYPE(self) : self;
+        tuple=PyTuple_Pack(2, self_type, param_type);
```
```suggestion
        tuple = PyTuple_Pack(2, self_type, param_type);
```
(Also below.)

gvanrossum commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,174 @@
+
+PyTypeObject Py_UnionType = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    .tp_name = "typing.Union",
```
```suggestion
    .tp_name = "types.Union",
```

gvanrossum replied:
Also you should probably add something to Lib/types.py along the lines of `Union = type(int|str)`.

gvanrossum commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,174 @@
+Py_Union_AddToTuple(PyObject *existing_param, PyObject *new_param, int addFirst)
+{
+    int position = 0;
+    PyObject* existingArgs = PyObject_GetAttrString((PyObject *)existing_param, "__args__");
```
Error check here and after `tuple = ...` below.

gvanrossum commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,174 @@
+PyObject *
+Py_Union_AddToTuple(PyObject *existing_param, PyObject *new_param, int addFirst)
+{
+    int position = 0;
```
If we assume that `addFirst` is either 0 or `, we won't need `position`.

gvanrossum commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,174 @@
+
+// TODO: MM - If we can somehow sort the arguments reliably, we wouldn't need the boolean parameter here.
+PyObject *
+Py_Union_AddToTuple(PyObject *existing_param, PyObject *new_param, int addFirst)
```
(Sorry for so many comments. :-)

I wonder if this couldn't be generalized to constructing a union of two unions as well -- i.e., `(int|str) | (float|complex)` should equal `int|str|float|complex`.

gvanrossum commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,174 @@
+};
+
+PyObject *
+Py_Union(PyObject *args)
```
This function and the next should both replicate the functionality of `_remove_dups_flatten()` from typing.py.

gvanrossum commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,174 @@
+        return PyObject_RichCompare(aa->args, b_args, Py_EQ);
+    }
+    PyTypeObject *type = Py_TYPE(b);
+    if (strcmp(type->tp_name, "typing.Union") != 0) {
```
The proper test just compares the pointers, i.e. `if (type == Py_UnionType)`

gvanrossum commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,174 @@
+};
+
+static PyObject *
+union_richcompare(PyObject *a, PyObject *b, int op)
```
This should presumably convert each of the args to a set and then compare the sets. (This way, we preserve the *visual* difference between `int|str` and `str|int`, but they still compare equal. Sort of like how two dicts with different item orders but the same sets of items print differently but compare equal.)

gvanrossum commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,174 @@
+    .tp_getattro = union_getattro,
+    .tp_members = union_members,
+    .tp_methods = union_methods,
+    .tp_richcompare = union_richcompare,
```
I recommend adding tp_repr. `repr(int|str)` should return the string `int|str`. (Similar to what genericalias.c does.)

gvanrossum commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,174 @@
+
+// TODO: MM: Implement this for isinstance checks
+static PyObject *
+union_instancecheck(PyObject *self, PyObject *instance)
```
I'll review this later.

gvanrossum reviewed (commented):
Great start! Here's a boatload of review comments, from whitespace nits to feature requests.

maggiemoss committed:
Add support for union use in subclasses.

gvanrossum commented on Lib/test/test_isinstance.py:
```diff
@@ -209,14 +209,14 @@ def test_isinstance_abstract(self):
-    #     self.assertEqual(False, isinstance(2, typing.List | typing.Tuple))
+    def test_isinstance_with_or_union(self):
+        # self.assertEqual(True, isinstance(AbstractChild(), AbstractChild | int))
+        self.assertEqual(False, isinstance(None, str | int))
```
How about adding a test that `isinstance(None, int|None)`?

gvanrossum commented on Objects/abstract.c:
```diff
@@ -2605,10 +2605,14 @@ recursive_issubclass(PyObject *derived, PyObject *cls)
+    int is_union = (PyType_Check(type) && strcmp(type->tp_name, "typing.Union") == 0);
+    if (!is_union && !check_class(cls,
+                            "issubclass() arg 2 must be a class"
+                            " or tuple of classes")) {
```
Weren't you going to add something to this error message?

gvanrossum commented on Objects/unionobject.c:
```diff
@@ -105,6 +122,8 @@ union_richcompare(PyObject *a, PyObject *b, int op)
 
 
+
+
```
These extra blank lines look suspect. :-)

gvanrossum reviewed (commented):
(Presumably you were going to add more commits before asking for another review?)

maggiemoss committed:
Fix code style inconsistencies.

maggiemoss committed:
Change is_not_union -> is_union

maggiemoss committed:
Implement dedup and flatten.

maggiemoss committed:
Add basic tp_repr to unionobject.

maggiemoss committed:
Update is_unionable.

maggiemoss committed:
Clean up whitespace.

maggiemoss committed:
Update test_types.

gvanrossum commented on Lib/typing.py:
```diff
@@ -117,6 +117,7 @@
 # namespace, but excluded from __all__ because they might stomp on
 # legitimate imports of those modules.
 
+Union = type(int|str)
```
This doesn't look right -- maybe you did this for testing?

gvanrossum commented on Objects/abstract.c:
```diff
@@ -2607,10 +2604,10 @@ recursive_issubclass(PyObject *derived, PyObject *cls)
     if (!is_union && !check_class(cls,
                             "issubclass() arg 2 must be a class"
-                            " or tuple of classes")) {
+                            " a tuple of classes, or a union.")) {
```
Aren't you missing a comma before "a tuple"?

gvanrossum commented on Objects/typeobject.c:
```diff
@@ -3741,85 +3741,10 @@ type_is_gc(PyTypeObject *type)
-    Py_DECREF(tuple);
-    return newUnionType;
+    PyObject* new_union = Py_Union_New(self, param);
+    return new_union;
```
You could even combine these two lines. :-)

gvanrossum commented on Objects/unionobject.c:
```diff
@@ -108,25 +108,282 @@ static PyMethodDef union_methods[] = {
 static PyObject *
 union_richcompare(PyObject *a, PyObject *b, int op)
 {
+    PyObject* b_set = NULL;
```
This probably should start by testing the value of `op`, and returning `NotImplemented` if it's not `EQ` or `NE`. Also for `NE` it should probably call itself with `op=EQ` and invert the result, or something. (Compare dict_richcompare() in dictobject.c.)

gvanrossum commented on Objects/unionobject.c:
```diff
@@ -108,25 +108,282 @@ static PyMethodDef union_methods[] = {
         PyObject* b_args = PyObject_GetAttrString(b, "__args__");
-        return PyObject_RichCompare(aa->args, b_args, Py_EQ);
+        b_set = PySet_New(b_args);
+
```
Drop this blank line.

gvanrossum commented on Objects/unionobject.c:
```diff
@@ -108,25 +108,282 @@ static PyMethodDef union_methods[] = {
+static int
+is_genericalias(PyObject *obj)
+{
+    return is_typing_name(obj, "GenericAlias");
```
This probably doesn't work, because `typing.GenericAlias` is just a coincidental alias for `types.GenericAlias`. (Check it in the REPL.)

gvanrossum reviewed (commented):
Huge progress! I have to admit I didn't double check every line, but I like how it's shaping up. Here is a smattering of review comments. I am going to have to contact the sponsor for PEP 604 to ensure this PEP will actually make a chance of appearing before the SC before 3.10 is too far underway.

maggiemoss committed:
Fix style and grammar issues.

maggiemoss committed:
Implement NE for union richcompare.

maggiemoss committed:
Remove is_generic_alias helper method.

gvanrossum commented on Lib/test/test_types.py:
```diff
@@ -643,6 +644,8 @@ def test_or_types_operator(self):
         with self.assertRaises(TypeError):
             Example() | int
+        with self.assertRaises(TypeError):
+            (int | str) < typing.Union[str, int]
```
I'd repeat this with `(int|str) < (str|int)`. Or even `<=`.

gvanrossum reviewed (commented):
Okay, now comes the hard part -- get all tests to pass...

maggiemoss committed:
Fix negative ref count.

maggiemoss committed:
Small code style fixes.

maggiemoss committed:
Add unionobject to pythoncore.vcxproj.filters

maggiemoss committed:
Remove semicolon after PyObject_HEAD

maggiemoss commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,433 @@
+{
+    args = flatten_args(args);
+    int arg_length = PyTuple_GET_SIZE(args);
+    PyObject* temp[arg_length];
```
This is causing issues in the windows build.

Is there a recommended workaround for variable length arrays? 

Alternatively, is there a maximum number of arguments one could have in a Union?

pablogsal replied:
These `args` will be subclasses of `type` no? Why not using a set (https://docs.python.org/3/c-api/set.html for the C-API of sets) to de-duplicate them (as subclasses of `type` will support `__hash__` and `__eq__`)?  Otherwise, this code will be `O(n**2)` on the number of types added.

gvanrossum replied:
I'd prefer not to use a set here, because we ought to return non-duplicate items in the order they came in. It matters because we want `repr(int | str)` to return `int | str`, not occasionally `str | int`.

I'm not worried about the _O(n**2)_ behavior here, these loops are just doing pointer comparisons and typical unions don't have that many members. But I don't want to declare a fixed maximum size. So I recommend `PyMem_Calloc()` and `PyMem_Free()`. (Alternatively, you could allocate a tuple of size `arg_length`, store the pointers there, and at the end use `_PyTuple_Resize()`. But it wouldn't look cleaner.)

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,433 @@
+
+    PyObject *param_type = param == Py_None ? (PyObject *)Py_TYPE(param) : param;
+    PyTypeObject *self_type = (PyObject *)self == Py_None ? Py_TYPE(self) : self;
+    PyObject *tuple = PyTuple_Pack(2, self_type, param_type);
```
You must check for the return value and handle errors

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,433 @@
+    PyObject *param_type = param == Py_None ? (PyObject *)Py_TYPE(param) : param;
+    PyTypeObject *self_type = (PyObject *)self == Py_None ? Py_TYPE(self) : self;
+    PyObject *tuple = PyTuple_Pack(2, self_type, param_type);
+    PyObject *new_union = Py_Union(tuple);
```
`Py_Union` will (should) return `NULL` on failure, so you need to handle that case

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,433 @@
+
+    // Create new tuple from deduplicated array.
+    PyObject* temp_arg = NULL;
+    PyObject* new_args = PyTuple_New(temp_count);
```
You need to check the return value and handle errors for `PyTuple_New`

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,433 @@
+        }
+    }
+    // Create new tuple of flattened args.
+    PyObject *flattened_args = PyTuple_New(total_args);
```
`PyTuple_New` can fail, you need to check the return code

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,433 @@
+static PyObject*
+dedup_and_flatten_args(PyObject* args)
+{
+    args = flatten_args(args);
```
`flatten_args` should return `NULL` on failure, so we need to check for the return code

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,433 @@
+        return NULL;
+    }
+
+    PyObject* new_args = dedup_and_flatten_args(args);
```
`dedup_and_flatten_args` should return `NULL` on failure, so a check is needed

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,433 @@
+
+    if (is_typing_name(b, "_UnionGenericAlias")) {
+        PyObject* b_args = PyObject_GetAttrString(b, "__args__");
+        b_set = PySet_New(b_args);
```
`PySet_New` can fail so a check is needed

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,433 @@
+
+    PyObject* b_set = NULL;
+    unionobject *aa = (unionobject *)a;
+    PyObject* a_set = PySet_New(aa->args);
```
`PySet_New` can fail so a check is needed

maggiemoss committed:
Use _PyTuple_Resize instead of variable length array.

maggiemoss committed:
Add error handling.

maggiemoss committed:
Add additional tests for union_richcompare.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

maggiemoss commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,435 @@
+        b_set = PySet_New(bb->args);
+    }
+
+    if (b_set == NULL || a_set == NULL) {
```
@pablogsal  they are both checked here.

pablogsal replied:
You should check them individually (or at least in a way that you don't try to construct `b_set` if `a_set` fails). This is because if `a_set` fails it will set an exception and if-then `b_set` fails it will override it or even crash (you cannot call Python code / C-API construction code with the error indicator set).

maggiemoss replied:
Ah, thanks for the explanation - I will update this check :)

maggiemoss committed:
Update error check for PySet_New call.

maggiemoss committed:
Allow Any and other _SpecialForm

maggiemoss committed:
Update repr to handle NoneType properly.

maggiemoss committed:
Add Union to types.py

maggiemoss committed:
Merge branch 'PEP605-implementation' of github.com:MaggieMoss/cpython into PEP605-implementation

gvanrossum commented on Lib/test/test_types.py:
```diff
@@ -694,6 +695,15 @@ def test_or_type_operator_with_NewType(self):
+        assert typing.Any | str == typing.Union[typing.Any, str]
+        assert typing.NoReturn | str == typing.Union[typing.NoReturn, str]
+        assert typing.Optional[int] | str == typing.Union[typing.Optional[int], str]
+        assert typing.Union[int, bool] | str == typing.Union[typing.Union[int, bool], str]
```
`typing.Union` flattens its args.
```suggestion
        assert typing.Union[int, bool] | str == typing.Union[int, bool, str]
```

gvanrossum commented on Lib/test/test_types.py:
```diff
@@ -694,6 +695,15 @@ def test_or_type_operator_with_NewType(self):
+        assert typing.Union[int, bool] | str == typing.Union[typing.Union[int, bool], str]
+
+    def test_or_type_repr(self):
+        assert repr(int | None) == "int | None"
```
Hm, I think this should be fixed at the `__args__` level -- `(int | None).__args__ == (int, None)`. At least that would be more consistent with PEP 585: `list[None].__args__ == (None,)`.

(OTOH I just discovered that `typing.Optional[int]` does put `NoneType` in `__args__`, so now I'm not sure. This is probably not specified anywhere.)

maggiemoss replied:
It looks like replacing `None` with `type(None)` is done explicitly in many cases. Example: https://github.com/python/cpython/blob/master/Lib/typing.py#L138

Do you happen to know why this was done?

gvanrossum replied:
Unfortunately I don't recall why. It has been there for a very long time. We didn't follow this for the new `GenericAlias` type though. If you just keep the `None` everywhere for the new `Union` type, does anything fail? I guess you would have to special-case it for `isinstance()` and `issubclass()` -- how ugly would that look?

Another approach might be to change `GenericAlias` to do the replacement of `None` with `type(None)`. (Then again, the new `GenericAlias` is intentionally pretty dumb -- you can write `list[42]` without getting an error. But for the new `Union` we'll have to have somewhat higher standards, due to the desire for `isinstance()` and `issubclass()` support (we intentionally disallow `GenericAlias` there).

maggiemoss replied:
Okay, so I've sorted out where I was going wrong here. 

In order to make sure that equality and isinstance work as we'd previously defined, I had to add some extra handling for the `None` type. Let me know if that's not going to work long term.

gvanrossum commented on Lib/typing.py:
```diff
@@ -116,7 +116,7 @@
 # namespace, but excluded from __all__ because they might stomp on
 # legitimate imports of those modules.
-
+# Union = type(int | str)
```
Why is this back (albeit commented out :-)?

maggiemoss replied:
Oof. 🙈  Merge issue, I suspect. Thanks for catching it :)

gvanrossum commented on Lib/test/test_types.py:
```diff
@@ -694,6 +695,15 @@ def test_or_type_operator_with_NewType(self):
+    def test_or_type_operator_with_SpecialForm(self):
+        assert typing.Any | str == typing.Union[typing.Any, str]
+        assert typing.NoReturn | str == typing.Union[typing.NoReturn, str]
+        assert typing.Optional[int] | str == typing.Union[typing.Optional[int], str]
```
(Which is equal to `typing.Union[int, NoneType, str]`.)

gvanrossum reviewed (commented):
Great progress with minimal new code! Although `int|int` still returns a `types.Union` instance rather than just the type `int` (which is displayed as `<class 'int'>`).

maggiemoss committed:
Remove commented out code.

maggiemoss committed:
Update test with nested unions.

maggiemoss committed:
Add additional tests for optionals.

maggiemoss committed:
Add support for int | int == int

maggiemoss committed:
Add more error handling to unionobject

maggiemoss committed:
Handle NoneType vs. None

maggiemoss committed:
Don't allow generics in isinstance calls.

maggiemoss committed:
Add errors for union subclass.

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+};
+
+static PyObject *
+union_getattro(PyObject *self, PyObject *name)
```
You can remove this function and place `PyObject_GenericGetAttr` directly on the slot:

```
    .tp_getattro = PyObject_GenericGetAttr,

```

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    Py_ssize_t nargs = PyTuple_GET_SIZE(alias->args);
+    for (Py_ssize_t iarg = 0; iarg < nargs; iarg++) {
+        PyObject *arg = PyTuple_GET_ITEM(alias->args, iarg);
+        PyTypeObject* arg_type =  Py_TYPE(arg);
```
Small style Nit:

```suggestion
        PyTypeObject *arg_type =  Py_TYPE(arg);
```

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+            return NULL;
+        }
+    }
+    return instance;
```
This seems incorrect: not only is not incrementing the reference count but I do not understand why we are returning `instance` here. Indeed, I can make the interpreter segfault with this check:

```
❯ ./python
Python 3.10.0a0 (heads/PEP605-implementation:9f014e204e, Aug 12 2020, 00:59:05)
[GCC 10.1.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> isinstance(3.123, str|int)
True
./Include/object.h:516: _Py_NegativeRefcount: Assertion failed: object has negative ref count
Enable tracemalloc to get the memory block allocation traceback

object address  : 0x7f95919dc280
object refcount : -1
object type     : 0x7f95919dc490
[1]    67499 segmentation fault (core dumped)  ./python
```

But even if you fix the reference count, this seems incorrect because `instance` will always be true:

```
>>> isinstance(3.14, int | str)
True
```

Probably you want to return `False` here.

I think we should add some extra negative cases to the `test_isinstance.py` file.

The correct thing is to return a boolean directly (`Py_RETURN_TRUE`/`Py_RETURN_FALSE`) here as the caller is going to immediately check for truth-ness:

https://github.com/python/cpython/blob/master/Objects/abstract.c#L2496

Returning `False`/`True` directly is also:

* A fast-path in `PyObject_IsTrue`
* You protect yourself against the (very unlikely case) that a type returns False in `PyObject_IsTrue`

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+        if (PyType_Check(arg)) {
+            if (PyObject_IsInstance(instance, arg) != 0)
+            {
+                return self;
```
See my other comment regarding the return value (same comments apply).

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+            {
+                return self;
+            }
+        } else if (arg_type == &Py_GenericAliasType) {
```
I think this is weird to check here because depending on the order of the types you will get the error or not. For instance, imagine that you have

`(int, str, GenericAlias, float)`

and you pass an integer. Then you will normally return because the previous check will succeed but if you pass a float or something else you will get the `GenericAlias` error.

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+            arg = (PyObject *)arg_type;
+        }
+        if (PyType_Check(arg)) {
+            if (PyObject_IsInstance(instance, arg) != 0)
```
I think you can join these two conditions once we address the other stuff in this function

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+        if (PyType_Check(arg)) {
+            if (PyType_IsSubtype((PyTypeObject *)instance, (PyTypeObject *)arg) != 0)
+            {
+                return self;
```
Same as in the other function, your caller expects a `bool`: https://github.com/python/cpython/blob/master/Objects/abstract.c#L2579 and you are not incrementing the reference count of the return value,

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+            return NULL;
+        }
+    }
+    return instance;
```
Same comment as in the previous function, you probably want to return `False` here as right now everything is `True`:

```
>>> issubclass(object, float|str)
True
>>> issubclass(dict, float|str)
True
```

maggiemoss replied:
Updated, and added these examples as test cases.

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+                return self;
+            }
+        } else if (Py_TYPE(arg) == &Py_GenericAliasType) {
+            PyErr_SetString(PyExc_TypeError,
```
Same comment as the previous function regarding the error and the order.

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    for (Py_ssize_t iarg = 0; iarg < nargs; iarg++) {
+        PyObject *arg = PyTuple_GET_ITEM(alias->args, iarg);
+
+        if (PyType_Check(arg)) {
```
Same comment as the previous function: you probably can join these two checks into one conditional.

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+
+    PyTypeObject *type = Py_TYPE(b);
+    if (is_typing_name(b, "_UnionGenericAlias")) {
+        PyObject* b_args = PyObject_GetAttrString(b, "__args__");
```
Check the return value for errors

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    if (is_typing_name(b, "_UnionGenericAlias")) {
+        PyObject* b_args = PyObject_GetAttrString(b, "__args__");
+        int b_arg_length = PyTuple_GET_SIZE(b_args);
+        for (int i = 0; i < b_arg_length; i++) {
```
```suggestion
        for (Py_ssize_t i = 0; i < b_arg_length; i++) {
```

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    }
+
+    if (b_set == NULL) {
+        Py_RETURN_FALSE;
```
This is incorrect, if `PySet_New` fails (and `b_set` is `NULL`) you need to mark an error by returning `NULL` not `False`

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+        unionobject *bb = (unionobject *)b;
+        b_set = PySet_New(bb->args);
+    } else {
+        PyObject * tup = PyTuple_Pack(1, b);
```
Check the return value and handle errors

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+        b_set = PySet_New(bb->args);
+    } else {
+        PyObject * tup = PyTuple_Pack(1, b);
+        b_set = PySet_New(tup);
```
You need to clean `tup` after you are done.

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+        for (int i = 0; i < b_arg_length; i++) {
+            PyObject* arg = PyTuple_GET_ITEM(b_args, i);
+            if (arg == (PyObject *)&_PyNone_Type) {
+                PyTuple_SET_ITEM(b_args, i, Py_None);
```
I am not sure if is a good idea to mutate the right-hand side object when checking for equality. This operation should be done at construction, no?

maggiemoss replied:
This is due to an issue with `NoneType` vs. `None` in the new union syntax and in `typing.Union`. See this discussion here: https://github.com/python/cpython/pull/21515#discussion_r463390909

If you have any suggestions for how to remedy that situation, I'm all ears :)

pablogsal replied:
> If you have any suggestions for how to remedy that situation, I'm all ears :)

Is not enough to do the substitution `None` -> `NoneType` at creation (or union) time? Why do we need to do this lazily? I may be missing something.

In any case I would prefer to interpose `None` when doing the comparison instead of mutating the right-hand side object, as comparisons should not have side effects.

maggiemoss replied:
Thanks for clarifying :) Sounds good to me.

maggiemoss replied:
Sorry, forgot to answer your question here:

> Is not enough to do the substitution None -> NoneType at creation (or union) time? Why do we need to do this lazily? I may be missing something.

We want 

`assert int | None == typing.Union[int, None]`

And the `__repr__` of `int | None` to be `int | None` instead of `int | NoneType`. If we convert it at creation time, we print `NoneType`. As well, the `__args__` should mimic those in the genericalias type:

```
>>> list[None].__args__
(None,)
>>> (int | None).__args__
(<class 'int'>, None)
```


The `None` argument is special cased for `typing.Union` here: https://github.com/python/cpython/blob/master/Lib/typing.py#L394-L395

Can see the issue with side effects though, I'll work on a different fix for the issue :)

pablogsal replied:
Oh, I understand why we were doing this here. Thanks for the clarification.

> In any case I would prefer to interpose None when doing the comparison instead of mutating the right-hand side object, as comparisons should not have side effects.

In that case, I think it would still be better to substitute the set creation as:
```
        b_set = PySet_New(b_args);
```

by creating an empty set and one by one adding the elements in the tuple, substituting `NoneType` when we encounter `None`. In this case, we achieve the same effect but we don't need to mutate the right-hand side. Notice that changing the right-hand side can be problematic:

```python
>>> import typing
>>> x = int | None
>>> y = typing.Union[int, None]
>>> my_dict = {y: 42}
>>> y in my_dict
True
>>> x == y
True
>>> y in my_dict
False
```

This is the kind of side-effects that are dangerous if you mutate a hashable object, even more dangerous if is the side effect of some apparently passive operation.

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+        unionobject *bb = (unionobject *)b;
+        b_set = PySet_New(bb->args);
+    } else {
+        PyObject * tup = PyTuple_Pack(1, b);
```
Are all these code paths exercised in the test suite?

maggiemoss replied:
Yes, they are. 👍

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    }
+    // Add unique elements to an array.
+    int temp_count = 0;
+    for(int i = 0; i < arg_length; i++) {
```
```suggestion
    for (Py_ssize_t  = 0; i < arg_length; i++) {
```

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    for(int i = 0; i < arg_length; i++) {
+        int is_duplicate = 0;
+        PyObject* i_tuple = PyTuple_GET_ITEM(args, i);
+        for(int j = i + 1; j < arg_length; j++) {
```
```suggestion
        for (Py_ssize_t j = i + 1; j < arg_length; j++) {
```

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+        PyObject* i_tuple = PyTuple_GET_ITEM(args, i);
+        for(int j = i + 1; j < arg_length; j++) {
+            PyObject* j_tuple = PyTuple_GET_ITEM(args, j);
+            if (i_tuple == j_tuple) is_duplicate = 1;
```
PEP 7: Add braces to the conditional

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+        return NULL;
+    }
+    // Add unique elements to an array.
+    int temp_count = 0;
```
Maybe renaming this to `added_items`?

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    }
+
+    _PyTuple_Resize(&new_args, temp_count);
+    Py_INCREF(new_args);
```
Is this correct? I think this is leaking `new_args` as tou already have a fresh tuple created in this function by calling `PyTuple_New`

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    if (!PyObject_IsInstance(obj, (PyObject *)&PyFunction_Type)) {
+        return 0;
+    }
+    PyObject *module = PyObject_GetAttrString(obj, "__module__");
```
This code is very similar to `is_typing_name`. Maybe there is an oportunity to refactor the common part or reuse this code in the other function.

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+        is_new_type(obj) ||
+        is_special_form(obj) ||
+        PyType_Check(obj) ||
+        (PyObject_IsInstance(obj, (PyObject*)&Py_GenericAliasType)) ||
```
All these `PyObject_IsInstance` can return an error (`-1`) and set an exception and you need to propagate that  to your callers (in this case `union_new`).

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    }
+    return (
+        is_typevar(obj) ||
+        is_new_type(obj) ||
```
You are returning `0` in error here but other functions in this big "or" condition will return `-1`.

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+union_new(PyTypeObject* self, PyObject* param)
+{
+    // Check param is a PyType or GenericAlias
+    if ((param == NULL) || !is_unionable((PyObject *)param) || !is_unionable((PyObject*)self))
```
You need to distinguish the error case from the failures in `is_unionable`

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    }
+    PyObject *new_union = Py_Union(tuple);
+    if (new_union == NULL) {
+        return NULL;
```
Here you will be leaking `tuple`

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+        return NULL;
+    }
+    Py_INCREF(new_union);
+    return new_union;
```
You are leaking `tuple` here

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    if (new_union == NULL) {
+        return NULL;
+    }
+    Py_INCREF(new_union);
```
Is this increment correct? You already have a fresh union object from `Py_Union` no?

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+};
+
+static int
+union_repr_item(_PyUnicodeWriter *writer, PyObject *p)
```
I found the repr of this class a bit odd because is very different from other `repr` of other types:

```
>>> str | int | float
str | int | float
```

The classes themselves  use angle brackets:

```
>>> str
<class 'str'>
>>> int
<class 'int'>
>>> float
<class 'float'>
```

And the closes type (`typing.Union`) shows the class:

```
>>> Union[str, float, int]
typing.Union[str, float, int]
```

maggiemoss replied:
This is a good point. I'm interested to know what you would expect to be printed here. 

Something like:

```
types.Union[str, int, float]

# or 

<class 'str'> | <class 'int'> | <class 'float'>
```

pablogsal replied:
IMHO this should return the same as `typing.Union` because this is what the class is claiming to be (it even has `typing.Union` in the name)

maggiemoss replied:
cc: @gvanrossum just wanted to get your insight here. I know you'd previously expressed some opinions on what this should display.

gvanrossum replied:
I actually like the PEP's `str | float | int` output. It matches what we do for PEP 585 types like `list[int]` (it prints literally that).

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    if (p == Py_Ellipsis) {
+        // The Ellipsis object
+        r = PyUnicode_FromString("...");
+        goto done;
```
Suggestion: maybe rename `done` to `exit`?

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+        goto done;
+    }
+
+    if (_PyObject_LookupAttrId(p, &PyId___origin__, &tmp) < 0) {
```
Seems that you are only checking that the attribute is there, so why not use something like `_PyObject_HasAttrId` ?

maggiemoss replied:
I borrowed heavily from the `repr` function in generic alias. It seems to me like we're also updating the `tmp` variable here, and using it later in the function. https://github.com/python/cpython/blob/master/Objects/genericaliasobject.c#L37

pablogsal replied:
> It seems to me like we're also updating the `tmp` variable here, and using it later in the function.

But you throw it away inmediatele:

```
    if (tmp != NULL) {
        Py_DECREF(tmp);
```

and then you substitute it for another one:

```
	        if (_PyObject_LookupAttrId(p, &PyId___args__, &tmp) < 0) {
            goto exit;
```

that then you throw away immediately:


```
        if (tmp != NULL) {
            Py_DECREF(tmp);
```

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+        if (i > 0) {
+            if (_PyUnicodeWriter_WriteASCIIString(&writer, " | ", 3) < 0) {
+                goto error;
+            }
```
I think you can join these two conditionals

pablogsal commented on Lib/test/test_types.py:
```diff
@@ -2,13 +2,18 @@
 
```
Running this test with the readleak checker shows that there are some reference leaks (some of them are indicated in my review but there may be more):

```
❯ ./python -m test test_types -R :
0:00:00 load avg: 1.10 Run tests sequentially
0:00:00 load avg: 1.10 [1/1] test_types
beginning 9 repetitions
123456789
.........
test_types leaked [1123, 1121, 1121, 1121] references, sum=4486
test_types leaked [336, 335, 335, 335] memory blocks, sum=1341
test_types failed

== Tests result: FAILURE ==

1 test failed:
    test_types

Total duration: 1.3 sec
Tests result: FAILURE
```

maggiemoss replied:
thanks for catching that, when I run `./python -m test test_types -R :` there are no leaks :)

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+Py_Union(PyObject *args)
+{
+    if (!PyTuple_Check(args)) {
+        args = PyTuple_Pack(1, args);
```
What types are valid in this code path? Are you expecting only instances of `Py_UnionType`? If so, it may be good to add some check for that and then emit an error message

maggiemoss replied:
@pablogsal The checking for what's allowed in a unionobject is in the `union_new` method (Called via `PyUnion_New`). Should this be refactored into this method?

pablogsal replied:
But users can call `Py_Union` without calling `union_new` no? Is part of the public C-API

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+
+    PyObject* new_args = dedup_and_flatten_args(args);
+    if (new_args == NULL) {
+        return NULL;
```
I think you would be leaking `args` here

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    }
+    alias->args = new_args;
+    _PyObject_GC_TRACK(alias);
+    return (PyObject *) alias;
```
You are leaking `args`

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,473 @@
+    .tp_dealloc = unionobject_dealloc,
+    .tp_alloc = PyType_GenericAlloc,
+    .tp_free = PyObject_GC_Del,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
```
Does this type really need GC support? Is basically an augmented tuple that is also immutable no?

maggiemoss committed:
Implement code review suggestions.

- Reorder checking valid arguments for isinstance and issubclass methods.
- Rename variables as suggested.
- Combine checks, where possible.

maggiemoss committed:
Increase code sharing between is_typing_name and is_new_type

maggiemoss committed:
Add tests, fix reference leaks.

maggiemoss committed:
Update tp_flags for unionobject.

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,493 @@
+    .tp_dealloc = unionobject_dealloc,
+    .tp_alloc = PyType_GenericAlloc,
+    .tp_free = PyObject_GC_Del,
+    .tp_flags = Py_TPFLAGS_DEFAULT,
```
When you remove the GC support you also need to remove as well `tp_traverse` (and `union_traverse`), all the calls to `_PyObject_GC_TRACK` and `_PyObject_GC_UNTRACK` and substitute the call to `PyObject_GC_New` (basically search for `GC_` and either remove or substitute accordingly). (Ping me if you have questions or doubts about these APIs)

maggiemoss replied:
I've been trying to get removing this flag to work. But can't seem to fix the resulting memory leaks. Do you have any advice for further debugging?

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,493 @@
+    {0}
+};
+
+static int*
```
You probably want to return an integer here instead of an integer pointer

maggiemoss committed:
Return int.

maggiemoss committed:
Fix NoneType comparison side effect.

maggiemoss committed:
Merge branch 'PEP605-implementation' of github.com:MaggieMoss/cpython into PEP605-implementation

maggiemoss committed:
Remove GC flag from unionobject.

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,497 @@
+
+PyTypeObject Py_UnionType = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    .tp_name = "types.Union",
```
Is this `types.Union` or `typing.Union`?

maggiemoss replied:
`types.Union`

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,497 @@
+    }
+    for (Py_ssize_t iarg = 0; iarg < nargs; iarg++) {
+        PyObject *arg = PyTuple_GET_ITEM(alias->args, iarg);
+        PyTypeObject *arg_type =  Py_TYPE(arg);
```
```suggestion
        PyTypeObject *arg_type = Py_TYPE(arg);
```

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,497 @@
+};
+
+static int
+check_args(PyObject *args) {
```
As this function is just checking for `Py_GenericAliasType` maybe we can have a better name. I am not the greatest with names but what about `check_args_for_genericalias` or `is_generic_alias_in_args`?

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,497 @@
+    if (module == NULL) {
+        return -1;
+    }
+    Py_DECREF(module);
```
This seems wrong, you only can do the `Py_DECREF` after you have finished using it (so the calls to `PyUnicode_Check` and  `_PyUnicode_EqualToASCIIString` can crash.

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,497 @@
+    unionobject *aa = (unionobject *)a;
+    PyObject* a_set = PySet_New(aa->args);
+    if (a_set == NULL) {
+        Py_RETURN_FALSE;
```
If the set fails, you need to return `NULL`, not `False`, to mark the error

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,497 @@
+    }
+
+    PyTypeObject *type = Py_TYPE(b);
+    if (is_typing_name(b, "_UnionGenericAlias")) {
```
You need to check if `is_typing_name` return `-1` because that's an error you need to propagate

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,497 @@
+            if (b_set == NULL) {
+                PyObject* args = PyTuple_Pack(1, arg);
+                if (args == NULL) {
+                    return NULL;
```
You would be potentially leaking `a_set` here if I am not mistaken

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,497 @@
+                b_set = PySet_New(args);
+                Py_DECREF(args);
+                if (b_set == NULL) {
+                    return NULL;
```
You would be potentially leaking `a_set` here if I am not mistaken

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,497 @@
+            } else {
+                int err = PySet_Add(b_set, arg);
+                if (err == -1) {
+                    return NULL;
```
`You would be potentially leaking `a_set` here if I am not mistaken`

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,497 @@
+    if (is_typing_name(b, "_UnionGenericAlias")) {
+        PyObject* b_args = PyObject_GetAttrString(b, "__args__");
+        if (b_args == NULL) {
+            return NULL;
```
You would be potentially leaking `a_set` here if I am not mistaken

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,497 @@
+        unionobject *bb = (unionobject *)b;
+        b_set = PySet_New(bb->args);
+        if (b_set == NULL) {
+            Py_DECREF(a_set);
```
I think it will be cleaner to have a `goto` target that cleans `a_set` and maybe `b_set` if non-NULL and returns `NULL` to avoid all this repetition (and the more repetition that should appear if we fix the previous cases)

maggiemoss replied:
I agree, much cleaner! Thanks for the suggestion :)

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,497 @@
+        b_set = PySet_New(tuple);
+        Py_DECREF(tuple);
+    }
+    if (b_set == NULL) {
```
Is possible to reach this condition from the previous if/else if we enter the if (and not the else)? I think we are returning `NULL` in all cases when `b_set` fails to be created. If this is true we should introduce this check in the previous else (the one that starts with `PyObject *tuple = PyTuple_Pack(1, b);`).

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,497 @@
+        return NULL;
+    }
+    // Check param is a PyType or GenericAlias
+    if (!is_unionable((PyObject *)param) || !is_unionable((PyObject*)self))
```
If `is_unionable` returns -1 that marks and error and you need to return `NULL` instead

maggiemoss committed:
Implement code review suggestions

- Use goto target to reduce code repetition.
- Fix style nits.
- Rename method check_args -> is_generic_alias_in_args.
- Fix potential reference leaks.
- Check for error result from is_unionable.
- Small refactor to repr method

maggiemoss committed:
Refactor and remove PyUnion_New

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,495 @@
+PyObject *
+Py_Union(PyObject *args)
+{
+    if (!PyTuple_Check(args)) {
```
I think this code path is not reached by anything in the test suite, so it may be better to just add an `assert` for `PyTuple_Check` here.

pablogsal replied:
I have done this in 8c06c1d

pablogsal commented on Include/unionobject.h:
```diff
@@ -0,0 +1,13 @@
+extern "C" {
+#endif
+
+PyAPI_FUNC(PyObject *) Py_Union(PyObject *);
```
@gvanrossum These functions need to be documented in the C-API docs (and a test should be added to test_capi), no? If this is not intended to be public, then we should add an underscore to the name. If they are public, what is the contract of valid arguments? Just tuples?

gvanrossum replied:
I'm not sure I see a use case for extension modules constructing unions yet, so I am fine with making it private. We can always change our mind later -- not so easily the other way around.

pablogsal replied:
Ok, 47a4cb3 makes the API private for the time being.

pablogsal committed:
General cleanup, fix reference leaks, add new test for possible crashes

pablogsal committed:
Add newline

pablogsal committed:
Make unionobject APIs private

pablogsal reviewed (approved):
This looks good to me!

Thanks for all the work and patience @MaggieMoss! Great job :)

@gvanrossum I won't land it myself in case you or Maggie want to take a last check.

pablogsal commented on Objects/unionobject.c:
```diff
@@ -0,0 +1,469 @@
+    PyObject *r = NULL;
+    int err;
+
+    if (p == Py_Ellipsis) {
```
One last question: is it possible to hold `Ellipsis` objects here?

Currently, that seems not possible:

```
>>> ... | int
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for |: 'ellipsis' and 'type'
```

gvanrossum replied:
But why would you want that? `typing.Union[..., int]` doesn't work either. And `...` isn't a type.

pablogsal replied:
Precisely, but this code is checking for the Ellipsis object being there which is not possible, no?

maggiemoss replied:
I can take this out :) Thanks for catching.

pablogsal committed:
Add minor test for GenericAlias

gvanrossum reviewed (approved):
I think it's good! Thanks so much Maggie for all your work.

maggiemoss committed:
Remove ellipsis from repr.

maggiemoss committed:
Remove commented out test code.
